STYLE.md

```diff
@@ -17,6 +17,15 @@
 
 * Exceptions are allowed and encouraged where appropriate. When using exceptions, do not add
   additional error handing that cannot possibly happen in the case an exception is thrown.
+* Do use exceptions for:
+  - Configuration ingestion error handling. Invalid configurations (dynamic and
+    static) should throw meaningful `EnvoyException`s, the configuration
+    ingestion code will catch these.
+  - To indicate constructor failure.
+* Do not use exceptions on the data path for general purpose error handling.
```

Comment:
A bit more context is here: https://github.com/envoyproxy/envoy/pull/4707#discussion_r226758649

We need clearer guide line for using exception on data path.

Reply:
The main area in which exception handling is used currently on the data path is codecs/parsing which tend to have very deep call stacks, making return error handling onerous. I don't think we should prevent people from doing this since I think it's a reasonable pattern. I do agree we should make things more clear in what we expect (in general I think we expect people to not use exceptions when the error call stack is only a single call deep for example), but I would generally clarify that Envoy core will not catch any exceptions on the data path, so any not caught exception will cause the server to crash.